### ISSUE TITLE
Bundle licenses from `raco pkg new` under `LICENSES` directory

### ISSUE THREAD

**Issue by sarna** [racket/racket#4608]
**Is your feature request related to a problem? Please describe.**
[Sourcehut](https://sr.ht/) automatically detects if you forgot to attach a license to your repository. However, it can't detect all naming variations, as described in [this thread](https://lists.sr.ht/~sircmpwn/sr.ht-discuss/%3C20210817090820.qny7tpd6ddfta2mn%40archdesk.localdomain%3E#%3C20210817103033.hteas4km24zovxmj@arch%3E). Racket's default dual licensing isn't detected and leaves all repos with a "We couldn't find a license file for your repository" warning.

**Describe the solution you'd like**
The best solution I found is bundling license files under a `LICENSES` directory, as mandated by the [REUSE specification](https://reuse.software/) (suggested in mailing thread mentioned above). Many projects follow best practices mandated by this specification, including projects like curl and Linux.

**Describe alternatives you've considered**
I considered pasting both licenses into one file or keeping only one of them - both demand manual work for each new repo, and have a less than satisfactory result.

**Do you want to contribute to this feature**
I can prepare a PR, though I would need pointers to where this part of `raco` lives - I wasn't able to find it myself.

**Comment by LiberalArtist** [comment]
My strongest opinion is that, if we do use a `LICENSES` directory, we should do it in a way that is actually consistent with the REUSE specification. That means the files should be `LICENSES/Apache-2.0.txt` and `LICENSES/MIT.txt`. Furthermore, I think it has implications for the files' contents. I just looked at the spec, and, while it defines "License File" as "a file containing the text of a license", it doesn't explicitly say what "the text of a license" is. My understanding, though, has always been that it should match according to [Annex B](https://spdx.github.io/spdx-spec/v2.3/license-matching-guidelines-and-templates/) of the SPDX specification, which uses the XML definition on the SPDX license list to account for missing or optional license elements. More conservatively, the files [`reuse download`](https://reuse.software/tutorial/#step-1) fetches are certainly acceptable. In any case, it should definitely be the full text of the `Apache-2.0` license, not:
https://github.com/racket/racket/blob/2c4e9908af2f68a60a5736c5dafdf05dd5f9bdb8/racket/collects/pkg/private/new.rkt#L95-L112

Another subtlety that would need to be addressed is that, since `raco pkg new foo` creates a single-collection package, creating a `LICENSES` directory would create a `foo/LICENSES` collection. Maybe this is just not a problem? Maybe the generated `info.rkt` file should define [`compile-omit-paths`](https://docs.racket-lang.org/raco/API_for_Raw_Compilation.html#%28idx._%28gentag._1._%28lib._scribblings%2Fraco%2Fraco..scrbl%29%29%29) or something? Or maybe it should supply them to something like [`copy-shared-files`](https://docs.racket-lang.org/raco/setup-info.html#%28idx._%28gentag._28._%28lib._scribblings%2Fraco%2Fraco..scrbl%29%29%29)? Even if we don't change `raco pkg new`, it would be good to know the right answers to those questions for packages that do use a `LICENSES` directory.

Does SourceHut document what it does detect, or do you know where the code they use for detection lives? For example, GitHub documents that they use the Ruby library [Licensee](https://github.com/licensee/licensee/blob/main/docs/what-we-look-at.md). (An extra complication is that Licensee apparently doesn't support a `LICENSES` directory at the moment, though in https://github.com/licensee/licensee/issues/490 it sounds like they'd be happy to receive a patch to add support.)

Overall, I'm neutral-to-positive about this proposal, providing the issues can be overcome.

**Comment by otherjoel** [comment]
Is there a reason why new packages should default to being dual-licensed?

**Comment by sorawee** [comment]
I think dual-licensing a reasonable default, since it mirrors Racket's own license. Of course, users are free to change it to whatever they want.

**Comment by LiberalArtist** [comment]
That's what was adopted when `raco pkg new` was changed from `LGPL-2.0-or-later` in https://github.com/racket/racket/pull/2701 after discussion in https://github.com/racket/racket/issues/1570#issuecomment-501123607 (buried in the enormous relicensing thread) with @sorawee, @samth, and @spdegabrielle.

I think the best reason to license new packages this way is because it's the license Racket as a project uses, so it makes sharing code as seamless as possible.

For another interpretation of "why", @samth wrote in the original post in https://github.com/racket/racket/issues/1570:

> We plan to switch to dual-license under the [Apache License, version 2](https://opensource.org/licenses/Apache-2.0) and the [MIT license](https://opensource.org/licenses/MIT). The MIT license is extremely permissive and compatible with almost everything, while the Apache license has patent protection and is compatible with other software already distributed under the Apache license. This is also the combination chosen by the Rust project.

In particular, Apache and other licenses with patent protection (good!) are generally thought (IANAL) to be incompatible with (L)GPL licenses before version 3.0, and there is some significant software that falls into that category. This was Rust's [reason](https://internals.rust-lang.org/t/rationale-of-apache-dual-licensing/8952/4) for dual-licensing rather than going for Apache only. An aspect I didn't understand until recently is how Apache continues to provide patent protection in this case: it's in the original Rust relicensing [email](https://mail.mozilla.org/pipermail/rust-dev/2012-November/002664.html), but [§ 3.4 of this technical report](https://doi.org/10.2172/1782513) (though I don't endorse all of its conclusions) summarizes it well:

> The dual license works because contributors must contribute code under the terms of *both* licenses, which means that they provide a patent license under `Apache-2.0`. Users, however, can use either `Apache-2.0` or `MIT`, and `MIT` is `GPL-2.0`-compatible, so `GPL-2.0` projects can use the code under `MIT`.

There are many other good options for licensing code, and I've written packages under many licenses. I would not have proposed `(Apache-2.0 OR MIT)` myself, either today or 2017, but I don't think we want to open up the can of worms of whether `(Apache-2.0 WITH LLVM-Exception)`, `MPL-2.0`, or anybody else's favorite license would have been a better choice.

**Comment by LiberalArtist** [comment]
Incidentally, @sorawee, I see you explicitly argued in https://github.com/racket/racket/issues/1570#issuecomment-501576060 that the contents of the current `LICENSE-APACHE` files should be what I argued above is wrong. In the [instructions](https://www.apache.org/licenses/LICENSE-2.0#apply) for applying the license that you cited, the first thing it says is to "include a copy of the Apache License, typically in a file called `LICENSE`, in your work"&emdash;which we don't, currently. Instead, `LICENSE-APACHE` currently consists of the "boilerplate declaration" that the recommend putting in a comment "to apply the Apache License to specific files in your work".

**Comment by otherjoel** [comment]
I understand the reasons these licenses were chosen for Racket itself (I was heavily involved with the relicensing effort at the end). But I don’t understand how code sharing for *new packages* would be harder if `raco pkg new` just defaulted to MIT.  Authors of new packages are not contributing to Racket, so in fact any permissive license would seem to work here just fine.

**Comment by LiberalArtist** [comment]
Yes, `MIT`-only wouldn't be a problem for code sharing. I would oppose recommending `MIT` alone as a default license because of the lack of patent protection. (And that's the sort of debate you get into if you make the license something other than "what Racket does".)

**Comment by sorawee** [comment]
OK, that was confusing. If I recall correctly, my understanding about how to apply the Apache License is really from [the license file itself](https://www.apache.org/licenses/LICENSE-2.0.txt):

```
   APPENDIX: How to apply the Apache License to your work.

      To apply the Apache License to your work, attach the following
      boilerplate notice, with the fields enclosed by brackets "[]"
      replaced with your own identifying information. (Don't include
      the brackets!)  The text should be enclosed in the appropriate
      comment syntax for the file format. We also recommend that a
      file or class name and description of purpose be included on the
      same "printed page" as the copyright notice for easier
      identification within third-party archives.

[THE BOILERPLATE]
```

And in an attempt to to link to this paragraph, I found https://www.apache.org/licenses/LICENSE-2.0#apply, whose text looks similar, so I assumed that they are exactly the same and linked it in the thread. Now I see that they are different. It also appears to actually talk about applying the license to a specific file, not the whole project. Do I read it correctly?

Looking at the current situation, I think that the instruction in https://www.apache.org/licenses/LICENSE-2.0#apply seems more elaborated, so I agree that we probably should follow that.

**Comment by spdegabrielle** [comment]
Is the proposed solution (LICENSES directory) supported by the repo hosts popular with Racket users? 

github, gitlab, sourcehut, others (?),

sorry if this has been addressed but I can’t find it in the thread above

**Comment by samth** [comment]
I haven't fully followed the discussion, but it sounds like this change would be better by SourceHut and worse for GitHub right now. Is that accurate? That doesn't seem like a change we would want in that case.

**Comment by LiberalArtist** [comment]
> I haven't fully followed the discussion, but it sounds like this change would be better by SourceHut and worse for GitHub right now. Is that accurate? That doesn't seem like a change we would want in that case.

Yes, that's the bottom line at this point, and I agree that we shouldn't break things that currently work. I still like this proposal in principle, but https://github.com/racket/racket/issues/4608#issuecomment-1483882075 would have to happen before this would be a viable option.

We can act now on the tangent from https://github.com/racket/racket/issues/4608#issuecomment-1483868969 and https://github.com/racket/racket/issues/4608#issuecomment-1483875002, though: I think @sorawee and I are agreed that we should replace the content of our `LICENSE-APACHE{,.txt}` files, including those generated by `raco pkg new`, with the full text of the license, i.e. <https://www.apache.org/licenses/LICENSE-2.0.txt>.
